src/Extension/CoreExtension.php

```diff
@@ -939,31 +939,100 @@ function twig_in_filter($value, $compare)
         $value = (string) $value;
     }
 
-    if (\is_array($compare)) {
-        return \in_array($value, $compare, \is_object($value) || \is_resource($value));
-    } elseif (\is_string($compare) && (\is_string($value) || \is_int($value) || \is_float($value))) {
-        return '' === $value || false !== strpos($compare, (string) $value);
-    } elseif ($compare instanceof \Traversable) {
-        if (\is_object($value) || \is_resource($value)) {
-            foreach ($compare as $item) {
-                if ($item === $value) {
-                    return true;
-                }
-            }
-        } else {
-            foreach ($compare as $item) {
-                if ($item == $value) {
-                    return true;
-                }
-            }
+    if (\is_string($compare)) {
+        if (\is_string($value) || \is_int($value) || \is_float($value)) {
+            return '' === $value || false !== strpos($compare, (string) $value);
         }
 
         return false;
     }
 
+    if (\is_object($value) || \is_resource($value)) {
+        return in_array($value, $compare, true);
+    }
+
+    if (!\is_array($compare) && !$compare instanceof \Traversable) {
+        return false;
+    }
+
+    foreach ($compare as $item) {
+        if (twig_compare($value, $item)) {
+            return true;
+        }
+    }
+
     return false;
 }
 
+/**
+ * Compares two values using a more strict version of the PHP non-strict comparison operator.
+ *
+ * @see https://wiki.php.net/rfc/string_to_number_comparison
+ * @see https://wiki.php.net/rfc/trailing_whitespace_numerics
+ *
+ * @internal
+ */
+function twig_compare($a, $b)
+{
+    // strict?
+    if (\is_object($a) || \is_resource($a) || \is_object($b) || \is_resource($b)) {
+        if ($a instanceof \SimpleXMLElement || $b instanceof \SimpleXMLElement) {
```

Comment:
Maybe instances of `\DateTimeInterface` should be compared with `==` to, it would be deceptive if dates are equals only if they are the same instance.

Reply:
the whole condition has been removed now.